modules/tensor_mechanics/src/materials/cohesive_zone_model/CZMComputeDisplacementJumpTotalLagrangian.C

```diff
@@ -0,0 +1,114 @@
+    _czm_total_rotation(declarePropertyByName<RankTwoTensor>(_base_name + "czm_total_rotation")),
+    _czm_total_rotation_inc(
+        declarePropertyByName<RankTwoTensor>(_base_name + "czm_total_rotation_inc"))
+{
+}
+
+void
+CZMComputeDisplacementJumpTotalLagrangian::initialSetup()
+{
+  // intialize displacementes vectors
+  CZMComputeDisplacementJumpBase::initialSetup();
+
+  // initializing the displacement gradeint vectors
+  for (unsigned int i = 0; i < _ndisp; ++i)
+  {
+    _grad_disp.push_back(&coupledGradient("displacements", i));
+    _grad_disp_neighbor.push_back(&coupledNeighborGradient("displacements", i));
+  }
+
+  // All others zero (so this will work naturally for 2D and 1D problems)
+  for (unsigned int i = _ndisp; i < 3; i++)
+  {
+    _grad_disp.push_back(&_grad_zero);
+    _grad_disp_neighbor.push_back(&_grad_zero);
+  }
+}
+
+void
+CZMComputeDisplacementJumpTotalLagrangian::initQpStatefulProperties()
+{
+  CZMComputeDisplacementJumpBase::initQpStatefulProperties();
+  _displacement_jump_global[_qp] = 0;
+  _F[_qp] = RankTwoTensor::Identity();
+  _R[_qp] = RankTwoTensor::Identity();
+}
+
+void
+CZMComputeDisplacementJumpTotalLagrangian::computeLocalDisplacementJump()
+{
+  computeFandR();
+  _czm_total_rotation_inc[_qp] = (_R[_qp] - _R_old[_qp]) * _czm_reference_rotation[_qp];
+  _czm_total_rotation[_qp] = _R[_qp] * _czm_reference_rotation[_qp];
+  _displacement_jump_global_inc[_qp] =
+      _displacement_jump_global[_qp] - _displacement_jump_global_old[_qp];
+
+  _interface_displacement_jump_inc[_qp] =
+      _czm_total_rotation_inc[_qp].transpose() * _displacement_jump_global[_qp] +
+      _czm_total_rotation[_qp].transpose() * _displacement_jump_global_inc[_qp];
+  _interface_displacement_jump[_qp] =
+      _interface_displacement_jump_old[_qp] + _interface_displacement_jump_inc[_qp];
+}
+
+void
+CZMComputeDisplacementJumpTotalLagrangian::computeFandR()
+{
+  RankTwoTensor F =
+      (RankTwoTensor::Identity() +
+       RankTwoTensor((*_grad_disp[0])[_qp], (*_grad_disp[1])[_qp], (*_grad_disp[2])[_qp]));
+  RankTwoTensor F_neighbor =
+      (RankTwoTensor::Identity() + RankTwoTensor((*_grad_disp_neighbor[0])[_qp],
+                                                 (*_grad_disp_neighbor[1])[_qp],
+                                                 (*_grad_disp_neighbor[2])[_qp]));
+
+  _F[_qp] = 0.5 * (F + F_neighbor);
+  // According to Cody mooseError are always fatal, so nothing we can do about
+  // them. The norm of the tensor might work, but there is the risk of an
+  // unwanted overflow (I tried and it happens). So checking component by
+  // component remains the only reasonable strategy. If someone finds a  better
+  // way this could be changed in the future
+  for (uint i = 0; i < 3; i++)
+    for (uint j = 0; j < 3; j++)
+      if (!std::isfinite(_F[_qp](i, j)))
+        throw MooseException("CZMMaterialBaseIncremental _F is not finite, reducing time step");
```

Comment:
What is the consequence of F having infinite value? Would solver terminate? Or it can automatically cut time step?

Reply:
nonfinite values trigger termination, so this is a safeguard to reduce tilmestep before making the all simulation crash. I have't found any other way around it, and in CP simulations sometimes it happens